File: src_all/libs_src/MobileIMSDK4i_x_open/MobileIMSDK4i/Core/Impl/AutoReLoginDaemon.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_X (MobileIMSDK v3.x) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址: http://www.52im.net/forum-89-1.html
//  > 即时通讯技术社区：http://www.52im.net/
//  > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//
//  如需联系作者，请发邮件至 jack.jiang@52im.net 或 jb2011@163.com.
//  ----------------------------------------------------------------------
//
//  AutoReLoginDaemon.m
//  MibileIMSDK4i_X (MobileIMSDK v3.0 at Summer 2017)
//
//  Created by JackJiang on 14/10/24.
//  Copyright (c) 2017年 52im.net. All rights reserved.
//

#import "AutoReLoginDaemon.h"
#import "ClientCoreSDK.h"
#import "LocalUDPDataSender.h"
#import "LocalUDPDataReciever.h"

static int AUTO_RE_LOGIN_INTERVAL = 2000;


@interface AutoReLoginDaemon ()

@property (nonatomic, assign) BOOL autoReLoginRunning;
@property (nonatomic, assign) BOOL _excuting;
@property (nonatomic, retain) NSTimer *timer;
@property (nonatomic, copy) ObserverCompletion debugObserver_;// block代码块一定要用copy属性，否则报错！

@end


@implementation AutoReLoginDaemon

static AutoReLoginDaemon *instance = nil;

+ (AutoReLoginDaemon *)sharedInstance
{
    if (instance == nil)
    {
        instance = [[super allocWithZone:NULL] init];
    }
    return instance;
}

+ (void) setAUTO_RE_LOGIN_INTERVAL:(int)autoReLoginInterval
{
    AUTO_RE_LOGIN_INTERVAL = autoReLoginInterval;
}
+ (int) getAUTO_RE_LOGIN_INTERVAL
{
    return AUTO_RE_LOGIN_INTERVAL;
}


- (id)init
{
    if (![super init])
        return nil;
    
    NSLog(@"AutoReLoginDaemon已经init了！");
    
    self.autoReLoginRunning = NO;
    self._excuting = NO;
    
    return self;
}

- (void) run
{
    if(!self._excuting)
    {
        self._excuting = YES;
        if([ClientCoreSDK isENABLED_DEBUG])
            NSLog(@"【IMCORE】自动重新登陆线程执行中, autoReLogin? %d...", [ClientCoreSDK isAutoReLogin]);
        int code = -1;

        if([ClientCoreSDK isAutoReLogin])
        {
            NSString *curLoginUserId = [ClientCoreSDK sharedInstance].currentLoginUserId;
            NSString *curLoginToken = [ClientCoreSDK sharedInstance].currentLoginToken;
            NSString *curLoginExtra = [ClientCoreSDK sharedInstance].currentLoginExtra;
            code = [[LocalUDPDataSender sharedInstance] sendLogin:curLoginUserId withToken:curLoginToken andExtra:curLoginExtra];
            
            // form DEBUG
            if(self.debugObserver_ != nil)
                self.debugObserver_(nil, [NSNumber numberWithInt:2]);
        }
        
        if(code == 0)
        {
            if([ClientCoreSDK isENABLED_DEBUG])
                 NSLog(@"【IMCORE】自动重新登陆数据包已发出(iOS上无需自已启动UDP接收线程, GCDAsyncUDPTask自行解决了).");
        }
        
        //
        self._excuting = NO;
    }
}


- (void) stop
{
    if(self.timer != nil)
    {
        if([self.timer isValid])
            [self.timer invalidate];
        
        self.timer = nil;
    }
    self.autoReLoginRunning = NO;
    
    // form DEBUG
    if(self.debugObserver_ != nil)
        self.debugObserver_(nil, [NSNumber numberWithInt:0]);
}

- (void) start:(BOOL)immediately
{
    [self stop];
    
    self.timer = [NSTimer scheduledTimerWithTimeInterval:AUTO_RE_LOGIN_INTERVAL / 1000
                                                  target:self
                                                selector:@selector(run)
                                                userInfo:nil
                                                 repeats:YES];
    if(immediately)
        [self.timer fire];
    self.autoReLoginRunning = YES;
    
    // form DEBUG
    if(self.debugObserver_ != nil)
        self.debugObserver_(nil, [NSNumber numberWithInt:1]);
}

- (BOOL) isAutoReLoginRunning
{
    return self.autoReLoginRunning;
}

- (void) setDebugObserver:(ObserverCompletion)debugObserver
{
    self.debugObserver_ = debugObserver;
}

@end

File: sdk_src/TCP_Client/SwiftMobileIMSDK(ios端swift版,由社区网友开发,仅作学习参考)/SwiftMobileIMSDK/Transfer/AutoReLoginDaemon.swift
//
//  AutoReLoginDaemon.swift
//  SwiftMobileIMSDK
//
//  Created by fishbay on 2021/8/2.

//  与服务端通信中断后的自动登陆（重连）独立线程

import Foundation

class AutoReLoginDaemon {
    
    /// 自动重新登陆时间间隔（单位：毫秒），默认2000毫秒
    static var autoReLoginInterval: Int = 2000
    
    /// 当前心跳线程是否正在执行中
    var autoReLoginRunning: Bool = false
    
    var executing: Bool = false
    var timer: Timer?
    
    /// 本属性仅作DEBUG之用：DEBUG事件观察者
    var debugObserver: ObserverCompletion?
    
    /// 单例
    private static let instance = AutoReLoginDaemon()
    static func sharedInstance() -> AutoReLoginDaemon {
        return instance
    }
    /// 禁止外部通过构造方法创建对象
    private init() {
        CAPrint("AutoReLoginDaemon已经init了")
    }
    
    /// 设置自动重新登陆的间隔时间
    /// - Parameter autoReLoginInterval: 间隔时间
    static func setAutoReLoginInterval(autoReLoginInterval: Int) {
        Self.autoReLoginInterval = autoReLoginInterval
    }
    static func getAutoReLoginInterval() -> Int {
        return autoReLoginInterval
    }
    
    @objc func run() {
        if self.executing {
            return
        }
        
        self.executing = true
        
        let isAutoReLogin = ClientCoreSDK.isAutoReLogin()
        let isSocketReady = LocalSocketProvider.sharedInstance().isLocalSocketReady()
        
        if ClientCoreSDK.isEnableDebug() {
            CAPrint("【IMCORE-TCP】自动重新登陆线程执行中, autoReLogin ? \(isAutoReLogin), socketReady ? \(isSocketReady) ...")
        }
        
        var code: ErrorCode = .unknown
        
        // 是否允许自动重新登陆哦
        if isAutoReLogin {
            code = LocalDataSender.sharedInstance().sendLogin(loginInfo: ClientCoreSDK.sharedInstance().getCurrentLoginInfo())
            
            // for DEBUG
            self.debugObserver?(nil, 2)
        }
        
        if code == .commonCodeOK {
            if ClientCoreSDK.isEnableDebug() {
                CAPrint("【IMCORE-TCP】自动重新登陆数据包已发出")
            }
        }
        
        self.executing = false
    }
    
    func stop() {
        if self.timer != nil {
            if self.timer!.isValid {
                self.timer!.invalidate()
            }
            
            self.timer = nil
        }
        
        self.autoReLoginRunning = false
        
        // for DEBUG
        self.debugObserver?(nil, 0)
    }
    
    func start(immediately: Bool) {
        self.stop()
        
        // 执行延迟的单位是秒
        self.timer = Timer.scheduledTimer(timeInterval: TimeInterval(Self.autoReLoginInterval / 1000),
                                          target: self,
                                          selector: #selector(run),
                                          userInfo: nil,
                                          repeats: true)
        // 如果需要立即执行
        if immediately {
            self.timer?.fire()
        }
        
        self.autoReLoginRunning = true
        
        // for DEBUG
        self.debugObserver?(nil, 1)

    }
    
    /// 设置debug回调
    func setDebugObserver(debugObserver: @escaping ObserverCompletion) {
        self.debugObserver = debugObserver
    }
    
    func isAutoReLoginRunning() -> Bool {
        return self.autoReLoginRunning
    }
    
}
